Add ParticleInterfaceBase to set the particle world index for the particle plugins.

MFraters opened the pull request:
This pull request adds a ParticleInterfaceBase to set the particle world index in the particle plugins as suggested by @gassmoeller. Let me know if this was what you had in mind.

This pull request is parallel to #5920 and should be able to be review and merged independently.

tjhei commented on include/aspect/particle/interface.h:
```diff
@@ -0,0 +1,48 @@
+  namespace Particle
+  {
+    /**
+     * An abstract base class defining methods
```
Can you explain that this is the base class for all plugins related to particles including properties, generators, etc.? What is the purpose?

tjhei commented on include/aspect/particle/interface.h:
```diff
@@ -0,0 +1,48 @@
+    class ParticleInterfaceBase : public Plugins::InterfaceBase
+    {
+      public:
+        void set_particle_world(unsigned int particle_world_index);
```
please document

tjhei commented on include/aspect/particle/interface.h:
```diff
@@ -0,0 +1,48 @@
+        void set_particle_world(unsigned int particle_world_index);
+
+      private:
+        unsigned int particle_world_index;
```
please document what this stores.

tjhei commented on include/aspect/particle/property/interface.h:
```diff
@@ -733,6 +734,12 @@ namespace aspect
           parse_parameters (ParameterHandler &prm);
 
+          /**
+           * @brief Set the particle wrold for all particle properties
```
```suggestion
           * @brief Set the particle world for all particle properties
```

tjhei commented:
What is your motivation for storing the index instead of pointer?

MFraters commented:
Thanks for the review! I have addressed your comments.

> What is your motivation for storing the index instead of pointer?

We need the index for the parse parameters, so since indices here used anyway, it made most sense to me to store it that way. I could change it to a pointer, but that would turn the interface into a template<int dim>, and  we would lose the information on what index the plugin is if we ever need the index during the calculation. Is there an advantage to having it as a pointer?

gassmoeller commented on include/aspect/particle/interface.h:
```diff
@@ -0,0 +1,59 @@
+         * Stores the index to the particle world in the
+         * particle worlds vector, to which the plugin belongs.
+         */
+        unsigned int particle_world_index;
```
Yes, except I do not see how you can access this index now. Either you need to introduce a `get_particle_world` function, or you need to make this variable `protected`.

gassmoeller commented on include/aspect/particle/interpolator/interface.h:
```diff
@@ -21,6 +21,7 @@
 #define _aspect_particle_interpolator_interface_h
 
+#include <aspect/particle/interface.h>
 #include <aspect/plugins.h>
```
no need to keep `plugins.h` now. Here and in the other interface headers.

gassmoeller reviewed (commented):
I think this goes in the right direction. It is too late for a proper review right now, but I also think storing the index is the right approach. Every plugin that reads parameters will have to know the world index (to construct the subsection it reads the parameters from), but only very few actually need a pointer to the particle world itself (and they can get it through simulator access).

MFraters commented:
@gassmoeller Thanks for your initial comments, I have addressed them.

gassmoeller commented on include/aspect/particle/interface.h:
```diff
@@ -0,0 +1,64 @@
+#ifndef _aspect_particle_interface_h
+#define _aspect_particle_interface_h
+
+#include "aspect/plugins.h"
```
```suggestion
#include <aspect/plugins.h>
```

gassmoeller commented on include/aspect/particle/interface.h:
```diff
@@ -0,0 +1,64 @@
+  namespace Particle
+  {
+    /**
+     * An abstract base class defining methods and member variables
```
This class is not abstract (https://en.cppreference.com/w/cpp/language/abstract_class)
```suggestion
     * A base class defining methods and member variables
```

gassmoeller commented on include/aspect/particle/interface.h:
```diff
@@ -0,0 +1,64 @@
+    {
+      public:
+        /**
+         * @brief Set which particle world the plugin belong to.
```
```suggestion
         * @brief Set which particle world the plugin belongs to.
```

gassmoeller commented on include/aspect/particle/interface.h:
```diff
@@ -0,0 +1,64 @@
+
+      private:
+        /**
+         * Stores the index to the particle world in the
+         * particle worlds vector, to which the plugin belongs.
```
Either call it the list of particle worlds, or like this:
```suggestion
         * Stores the index to the particle world, to which the plugin belongs.
```
No one should need to know how the particle worlds are stored.

gassmoeller commented on include/aspect/particle/property/interface.h:
```diff
@@ -733,6 +733,12 @@ namespace aspect
           parse_parameters (ParameterHandler &prm);
 
+          /**
+           * @brief Set the particle world for all particle properties
```
```suggestion
           * @brief Set the particle world index for all particle properties
```

gassmoeller commented on source/particle/interface.cc:
```diff
@@ -0,0 +1,41 @@
+/*
+  Copyright (C) 2015 - 2022 by the authors of the ASPECT code.
```
```suggestion
  Copyright (C) 2024 by the authors of the ASPECT code.
```

gassmoeller commented on source/particle/interface.cc:
```diff
@@ -0,0 +1,41 @@
+    {
+      this->particle_world_index = particle_world_index;
+    }
+
```
3 empty lines

gassmoeller commented on source/particle/interface.cc:
```diff
@@ -0,0 +1,41 @@
+    unsigned int
+    ParticleInterfaceBase::get_particle_world_index() const
+    {
+      return this->particle_world_index;
```
```suggestion
      return particle_world_index;
```

gassmoeller commented on source/particle/interface.cc:
```diff
@@ -0,0 +1,41 @@
+    {
+      return this->particle_world_index;
+    }
+
```
```suggestion
```

gassmoeller commented on source/particle/world.cc:
```diff
@@ -1333,27 +1333,31 @@ namespace aspect
           sim->initialize_simulator (this->get_simulator());
         generator->parse_parameters(prm);
         generator->initialize();
+        generator->set_particle_world_index(world_index);
```
This is the wrong order, the `parse_parameters` function will already need to know this index. Here and for the other functions.

gassmoeller reviewed (commented):
Yes, pretty good. My remaining comments are small, afterwards this can be merged.

MFraters commented:
@gassmoeller Thanks for the review and good catch for the order of the parse parameters. I have addressed your comments.
